lms/services/digest.py

```diff
@@ -1,11 +1,303 @@
+        unified_courses = []
+        for unified_course in self.unified_courses.values():
+            if unified_course not in unified_courses:
+                unified_courses.append(unified_course)
+
+        for unified_course in unified_courses:
+            num_annotations = len(unified_course.learner_annotations)
+
+            if not num_annotations:
+                # There was no activity in this course.
+                continue
+
+            if h_userid not in [user.h_userid for user in unified_course.instructors]:
+                # The user isn't an instructor in this course.
+                continue
+
+            course_digests.append(
+                {"title": unified_course.title, "num_annotations": num_annotations}
+            )
+
+        return {
+            "total_annotations": sum(
+                course_digest["num_annotations"] for course_digest in course_digests
+            ),
+            "courses": course_digests,
+        }
+
+    @property
+    def unified_users(self):
+        """Return a dict mapping h_userid's to UnifiedUser's."""
+        if self._unified_users is not None:
+            return self._unified_users
+
+        self._unified_users = {}
+
+        h_userids = set(
+            self._audience
+            + [annotation["author"]["userid"] for annotation in self._annotations]
+        )
+
+        for h_userid in h_userids:
+            users = self._db.scalars(
+                select(User).filter_by(h_userid=h_userid)
+                # Order most-recently-updated first so that the code below will
+                # use the most recent values for email, display name, etc.
+                .order_by(User.updated.desc())
+            ).all()
+
+            email = None
+            for user in users:
+                if user.email:
+                    email = user.email
+                    break
+
+            display_name = None
+            for user in users:
+                if user.display_name:
+                    display_name = user.display_name
+                    break
+
+            self._unified_users[h_userid] = UnifiedUser(
+                h_userid=h_userid,
+                users=tuple(users),
+                email=email,
+                display_name=display_name,
+            )
+
+        return self._unified_users
+
+    @property
+    def unified_courses(self):
+        """
+        Return a dict mapping authority_provided_id's to UnifiedCourse's.
+
+        Multiple keys in the dict may point to the same UnifiedCourse object,
+        meaning that unified_courses.values() may contain duplicates. This
+        happens when there are annotations in multiple sub-groupings that
+        belong to the same course.
+        """
+        if self._unified_courses is not None:
+            return self._unified_courses
+
+        self._unified_courses = {}
+
+        def is_instructor(unified_user: UnifiedUser, courses: Iterable[Course]) -> bool:
```

Comment:
Not a fan of inlining functions that are longer than a one liner, but just a style preference.

Reply:
I'll add a commit to move this and various other parts out into private helpers, it'll make it a lot more readablel